src/Microsoft.SymbolStore/ChecksumValidator.cs

```diff
@@ -0,0 +1,125 @@
+﻿using Microsoft.FileFormats.PE;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Security.Cryptography;
+using System.Text;
+
+namespace Microsoft.SymbolStore
+{
+    class ChecksumValidator
+    {
+        private const string pdbStreamName = "#Pdb";
+        private const uint pdbIdSize = 20;
+
+        internal static void Validate(ITracer tracer, Stream pdbStream, IEnumerable<PdbChecksum> pdbChecksums)
+        {
+            using (MemoryStream ms = new MemoryStream())
+            {
+                pdbStream.CopyTo(ms);
+                byte[] bytes = ms.ToArray();
+                uint offset = 0;
+
+                try
+                {
+                    offset = GetPdbStreamOffset(pdbStream);
+                }
+                catch (Exception ex)
+                {
+                    tracer.Error(ex.Message);
+                    throw;
+                }
+
+                for (int i = 0; i <= pdbIdSize; i++)
+                {
+                    bytes[i + offset] = 0;
+                }
+
+                foreach (var checksum in pdbChecksums)
+                {
+                    tracer.Information($"Testing checksum: {checksum}");
+
+                    var algorithm = HashAlgorithm.Create(checksum.AlgorithmName);
+                    if (algorithm != null)
+                    {
+                        var hash = algorithm.ComputeHash(bytes);
+                        if (hash.SequenceEqual(checksum.Checksum))
+                        {
+                            // If any of the checksums are OK, we're good
+                            tracer.Information($"Found checksum match {checksum}");
+                            return;
+                        }
+                    }
+                    else
```

Comment:
You don't want to throw in this case unless all the algorithm names are unknown -- this might be a new checksum that this framework doesn't support, but a newer framework does.

Reply:
Also, it should probably still throw an InvalidChecksumException, but with a different message.